registry/storage/cache/provider/cacheprovider.go

```diff
@@ -0,0 +1,39 @@
+package cacheprovider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/distribution/distribution/v3/registry/storage/cache"
+)
+
+// InitFunc is the type of a CacheProvider factory function and is
+// used to register the constructor for different CacheProvider backends.
+type InitFunc func(ctx context.Context, options map[string]interface{}) (cache.BlobDescriptorCacheProvider, error)
+
+var cacheProviders map[string]InitFunc
+
+// Register is used to register an InitFunc for
+// a CacheProvider backend with the given name.
+func Register(name string, initFunc InitFunc) error {
+	if cacheProviders == nil {
+		cacheProviders = make(map[string]InitFunc)
+	}
+	if _, exists := cacheProviders[name]; exists {
+		return fmt.Errorf("name already registered: %s", name)
+	}
```

Comment:
Is `provider.Register` intended to be called from `init()` functions for packages to self-register? If so, why not just panic?

Reply:
Good shout. I think there are a few other cases in this codebase where we should rectify this e.g. https://github.com/distribution/distribution/blob/04e3bdaa7c6304d14970f13c01f0a8ded8289e8d/registry/storage/driver/middleware/storagemiddleware.go#L18-L29